### DISCUSSION TITLE
Sonoff NSPanel Touch Display

### DISCUSSION THREAD

**Discussion by Mark-Muc** [arendst/Tasmota#17099] [Q&A]
Hello ...
I'd like to compile a w32 binary for Sonoff NSPanel Touch Display.
I first flashed it with https://tasmota.github.io/install/ using:
![image](https://user-images.githubusercontent.com/67410847/202190426-7fdcd781-b769-4f30-9460-41d759a9e47e.png)
 just worked fine .. applied the template
`{"NAME":"NSPanel","GPIO":[0,0,0,0,3872,0,0,0,0,0,32,0,0,0,0,225,0,480,224,1,0,0,0,33,0,0,0,0,0,0,0,0,0,0,4736,0],"FLAG":0,"BASE":1,"CMND":"ADCParam1 2,11200,10000,3950 | Sleep 0 | BuzzerPWM 1"}`
Alos worked fin .. like it should.
Now my question is:
with online compiler from `https://benzino77-tasmocompiler`
What parameter I have to choose to habe it in German language .. and is it possible to complied it with bluetooth and 
thermostat extension ..
Thx for help & answers!!
Mark-Muc

**Comment by Jason2866** [comment]
You have to set this defines
```
FIRMWARE_NSPANEL
FRAMEWORK_ARDUINO_ITEAD
```
the others should be check options in TasmoCompiler

**Reply by Mark-Muc** [reply]
@Jason2866: Thank you .. helped me a lot!!
One other question I have on this device:
I'd like to use the built in Temp-Sensor in the thermostat def.:
The compileswitches tell me:

```
#define USE_THERMOSTAT                           // Add support for Thermostat
    #define THERMOSTAT_CONTROLLER_OUTPUTS         1         // Number of outputs to be controlled independently
    #define THERMOSTAT_SENSOR_NAME                "**DS18B20**" // Name of the local sensor to be used
```
now I have to change `THERMOSTAT_SENSOR_NAME `
to??
UI tells me:
`Analog1 Temperatur	19.7 °C`
Definition in the Module says:
![image](https://user-images.githubusercontent.com/67410847/202671909-b72e8631-e812-4a5a-bf7d-cfeee3ca01cb.png)

so what to enter in:
  `  #define THERMOSTAT_SENSOR_NAME                "DS18B20" // Name of the local sensor to be used`
`"ADC Temp1"`  maybe??

Thx for any help!
Mark-Muc

**Comment by Jason2866** [comment]
UI info is irrelevant. Regarding this PR #8659 the mqtt string is parsed. So you have to use the name of the sensor which is used in mqtt message.

**Reply by Mark-Muc** [reply]
I'm stupid, I could have thought of that myself.
Mark-Muc

**Reply by Mark-Muc** [reply]
here are the compiler defines for building a version using the internal temp.
Regards 
Mark-Muc

```
//Sonoff NSPanel Touch Display
#define FIRMWARE_NSPANEL
#define FRAMEWORK_ARDUINO_ITEAD

//Thermostat
#define USE_THERMOSTAT                           // Add support for Thermostat
    #define THERMOSTAT_CONTROLLER_OUTPUTS         1         // Number of outputs to be controlled independently
//    #define THERMOSTAT_SENSOR_NAME                "DS18B20" // Name of the local sensor to be used <-- check MQTT!! "ANALOG"
//   MQTT = "ANALOG": {"Temperature1": 20.6}
    #define THERMOSTAT_SENSOR_NAME                "ANALOG" //internal name 

    #define THERMOSTAT_SENSOR_NUMBER              1         // Number of local sensors to be used
    #define THERMOSTAT_RELAY_NUMBER               1         // Default output relay number for the first controller (+i for following ones)
    #define THERMOSTAT_SWITCH_NUMBER              1         // Default input switch number for the first controller (+i for following ones)
    #define THERMOSTAT_TIME_ALLOW_RAMPUP          300       // Default time after last target update to allow ramp-up controller phase in minutes
    #define THERMOSTAT_TIME_RAMPUP_MAX            960       // Default time maximum ramp-up controller duration in minutes
    #define THERMOSTAT_TIME_RAMPUP_CYCLE          30        // Default time ramp-up cycle in minutes
    #define THERMOSTAT_TIME_SENS_LOST             30        // Maximum time w/o sensor update to set it as lost in minutes
    #define THERMOSTAT_TEMP_SENS_NUMBER           1         // Default temperature sensor number
    #define THERMOSTAT_TIME_MANUAL_TO_AUTO        60        // Default time without input switch active to change from manual to automatic in minutes
    #define THERMOSTAT_TIME_RESET                 12000     // Default reset time of the PI controller in seconds
    #define THERMOSTAT_TIME_PI_CYCLE              30        // Default cycle time for the thermostat controller in minutes
    #define THERMOSTAT_TIME_MAX_ACTION            20        // Default maximum thermostat time per cycle in minutes
    #define THERMOSTAT_TIME_MIN_ACTION            4         // Default minimum thermostat time per cycle in minutes
    #define THERMOSTAT_TIME_MIN_TURNOFF_ACTION    3         // Default minimum turnoff time in minutes, below it the thermostat will be held on
    #define THERMOSTAT_PROP_BAND                  4         // Default proportional band of the PI controller in degrees celsius
    #define THERMOSTAT_TEMP_RESET_ANTI_WINDUP     8         // Default range where reset antiwindup is disabled, in tenths of degrees celsius
    #define THERMOSTAT_TEMP_HYSTERESIS            1         // Default range hysteresis for temperature PI controller, in tenths of degrees celsius
    #define THERMOSTAT_TEMP_FROST_PROTECT         40        // Default minimum temperature for frost protection, in tenths of degrees celsius
    #define THERMOSTAT_TEMP_RAMPUP_DELTA_IN       4         // Default minimum delta temperature to target to get into rampup mode, in tenths of degrees celsius
    #define THERMOSTAT_TEMP_RAMPUP_DELTA_OUT      2         // Default minimum delta temperature to target to get out of the rampup mode, in tenths of degrees celsius
    #define THERMOSTAT_TEMP_PI_RAMPUP_ACC_E       200       // Default accumulated error when switching from ramp-up controller to PI in hundredths of degrees celsius
    #define THERMOSTAT_TIME_OUTPUT_DELAY          180       // Default output delay between state change and real actuation event (f.i. valve open/closed)
    #define THERMOSTAT_TEMP_INIT                  180       // Default init target temperature for the thermostat controller
    #define THERMOSTAT_TIME_MAX_OUTPUT_INCONSIST  3         // Default maximum time where the input and the output shall differ (for diagnostic) in minutes
    #define THERMOSTAT_TIME_MAX_AUTOTUNE          21600     // Maximum time for the PI autotune function to complete in seconds
    #define THERMOSTAT_DUTYCYCLE_AUTOTUNE         35        // Default duty cycle (in % over PI cycle time) for the step response of the autotune PI function
    #define THERMOSTAT_PEAKNUMBER_AUTOTUNE        8         // Default number of peak temperatures (max or min) to be used for the autotune PI function
    #define THERMOSTAT_TEMP_BAND_NO_PEAK_DET      1         // Default temperature band in thenths of degrees celsius within no peak will be detected
    #define THERMOSTAT_TIME_STD_DEV_PEAK_DET_OK   10    
```

**Reply by Mark-Muc** [reply]
Hello . one other stupid question.
I built a version matches the Sensorname from MQTT .see above!!
It seems to work .. checked by injection different `TempTargetSet xx.xx` 
but now the questen is: in the UI the  Current Temp & Gradient Temp are 0
![image](https://user-images.githubusercontent.com/67410847/202847016-2768b7a7-2364-4fff-b68c-13ab4a780cc2.png)

is it correct .. my thinking is .. ist should @ least match the Temperture shown above ..
`SensorInputSet 1` 
is set to 1 = Local sensor (can be changed by define, default DS18B20)
I think what's going wrong is here:
https://github.com/arendst/Tasmota/pull/8659/commits/440219fd91afef1b763a02efc80cadd9663d369d
![image](https://user-images.githubusercontent.com/67410847/202850920-92d9e0d2-4957-469a-8814-d758d1a72978.png)

it looks for a value in 
`const char* value_c = root[THERMOSTAT_SENSOR_NAME]["Temperature"];`
but the Sonoff NSPanel Touch Display MQTT reports:
`{"Time":"2022-11-19T13:31:49","ANALOG":{"Temperature1":20.9},"Thermostat0":{"ThermostatModeSet":1,"TempTargetSet":21.00,"CtrDutyCycleRead":66},"TempUnit":"C"}`

`Temperatur1 `instead of `Temperatur  `.. so the internal Analog Tempsensor can't be used anyhow...

TempMeasuredSet 
`13:36:09.576 MQT: stat/Thermo_WohnZi/RESULT = {"TempMeasuredSet1":0.0}`

Thx & Regards
Mark-Muc
maybe a simply "*" can help!
 ```
? - any character  (one and only one)
  * - any characters (zero or more)

```
